bin/propolis-server/src/lib/migrate/destination.rs

```diff
@@ -24,105 +27,321 @@ use crate::migrate::probes;
-        // Note, we don't use `?` here as this is a best effort and we don't
-        // want an error encountered during this send to shadow the run error
-        // from the caller.
-        if let Ok(e) = codec::Message::Error(err.clone()).try_into() {
-            let _ = proto.conn.send(e).await;
+    Ok(match selected {
+        Protocol::RonV0 => RonV0::new(log, migration_id, conn, local_addr),
+    })
+}
+
+/// A helper type for abstracting away type-level differences between the
+/// different phases of instance ensure.
+enum EnsureState<'a> {
+    NotStarted(VmEnsureNotStarted<'a>),
+    Active(VmEnsureActive<'a>),
+}
+
+impl<'a> EnsureState<'a> {
+    /// Moves the state machine to the `Active` state by creating VM objects,
+    /// preparing them for an incoming live migration, and activating the
+    /// created objects.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the state machine is not in the `NotStarted` state.
+    async fn activate(self) -> anyhow::Result<Self> {
+        let not_started = match self {
+            EnsureState::NotStarted(vm) => vm,
+            _ => unreachable!("can only activate a VM once"),
+        };
+
+        let mut objects_created = not_started.create_objects().await?;
+        objects_created.prepare_for_migration().await;
+        let active = objects_created.ensure_active().await;
+        Ok(Self::Active(active))
+    }
+
+    /// Publishes the supplied migration state to the server's external state
+    /// channel.
+    fn update(&mut self, id: Uuid, state: MigrationState) {
+        let publisher = match self {
+            EnsureState::NotStarted(vm) => vm.state_publisher(),
+            EnsureState::Active(vm) => vm.state_publisher(),
+        };
+
+        publisher.update(ExternalStateUpdate::Migration(
+            MigrationStateUpdate { state, id, role: MigrateRole::Destination },
+        ));
+    }
+
+    /// Notifies the instance ensure helper that migration has failed and that
+    /// VM initialization should therefore also fail.
+    async fn fail(self, reason: MigrateError) -> MigrateError {
+        match self {
+            EnsureState::NotStarted(vm) => {
+                let reason = vm.fail(reason.into()).await;
+                reason
+                    .downcast::<MigrateError>()
+                    .expect("original error was a MigrateError")
+            }
+            EnsureState::Active(vm) => {
+                vm.fail().await;
+                reason
+            }
         }
-        return Err(err);
     }
 
-    Ok(())
+    /// If the state machine is in the `Active` state, locks the VM objects
+    /// shared and returns the corresponding guard. Returns `None` otherwise.
+    async fn vm_objects(&self) -> Option<VmObjectsShared> {
+        let objects = match self {
+            EnsureState::NotStarted(_) => return None,
+            EnsureState::Active(vm) => vm.vm_objects(),
+        };
+
+        Some(objects.lock_shared().await)
```

Comment:
unimportant style nit, take it or leave it: this could be just
```suggestion
        match self {
            EnsureState::NotStarted(_) => None,
            EnsureState::Active(vm) => Some(vm.vm_objects().lock_shared().await),
        }
```
although if you think that's less clear, the current code is fine too...

Reply:
Done in 6943610